### PR TITLE
[SPARK-16675][SQL] Avoid per-record type dispatch in JDBC when writing

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

Currently, `JdbcUtils.savePartition` is doing type-based dispatch for each row to write appropriate values.

So, appropriate setters for `PreparedStatement` can be created first according to the schema, and then apply them to each row. This approach is similar with `CatalystWriteSupport`.

This PR simply make the setters to avoid this.
## How was this patch tested?

Existing tests should cover this.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Hi @cloud-fan , this one is about writing. Could you please take a look?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 1]
```diff
@@ -138,6 +138,80 @@ object JdbcUtils extends Logging {
       throw new IllegalArgumentException(s"Can't get JDBC type for ${dt.simpleString}"))
   }
 
+  // A `JDBCValueGetter` is responsible for converting, getting a value from `Row`
+  // and then setting into a field for `PreparedStatement`. The last argument
+  // `Int` means the index for the value to be set in the SQL statement and also used
+  // for the value to retrieve from `Row`.
+  private type JDBCValueGetter = (PreparedStatement, Row, Int) => Unit
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: hmm, I think `setter` is a more proper name here, and we should use `getter` for the read path.

`JDBCValueGetter`, sounds like we get the value from JDBC.

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Sure! thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 2]
```diff
@@ -138,6 +138,79 @@ object JdbcUtils extends Logging {
       throw new IllegalArgumentException(s"Can't get JDBC type for ${dt.simpleString}"))
   }
 
+  // A `JDBCValueSetter` is responsible for converting and setting a value from `Row` into
+  // a field for `PreparedStatement`. The last argument `Int` means the index for the
+  // value to be set in the SQL statement and also used for the value in `Row`.
+  private type JDBCValueSetter = (PreparedStatement, Row, Int) => Unit
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: please rename the read path too.

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Fixed!

**Comment by cloud-fan** [comment]
thanks, merging to master!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 3]
```diff
@@ -154,6 +154,79 @@ object JdbcUtils extends Logging {
...
+      (stmt: PreparedStatement, row: Row, pos: Int) =>
+        stmt.setDate(pos + 1, row.getAs[java.sql.Date](pos))
+
+    case t: DecimalType =>
+      (stmt: PreparedStatement, row: Row, pos: Int) =>
+        stmt.setBigDecimal(pos + 1, row.getDecimal(pos))
+
+    case ArrayType(et, _) =>
+      // remove type length parameters from end of type name
+      val typeName = getJdbcType(et, dialect).databaseTypeDefinition
```
Comment (jneira-stratio) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:216]: Late to the party :stuck_out_tongue_closed_eyes:  but why the type name is converted to lower case?
It makes code trying to get a JDBCType enum value using the string fails with: `java.lang.IllegalArgumentException: No enum constant java.sql.JDBCType.varchar` :disappointed:

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:216]: This is the same as the original code

Comment (jneira-stratio) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:216]: Oh yeah from https://github.com/apache/spark/pull/14323/files#diff-c3859e97335ead4b131263565c987d877bea0af3adbd6c5bf2d3716768d2e083L244
will continue digging to try to know why toLowerCase wad added, thanks!

**Pull request closed without merging**